apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+        Interval interval = bounds[i];
+        Expr extent = simplify(interval.max - interval.min + 1);
+        extent = simplify(substitute_var_estimates(extent));
+        const int64_t *extent_int = as_const_int(extent);
+        user_assert(extent_int != nullptr) <<
+            "extent:" << extent << " is not constant.\n";
+        int_bounds.push_back(*extent_int);
+    }
+    return int_bounds;
+}
+
+std::vector<int> get_rvar_bounds(const std::vector<ReductionVariable> &rvars) {
+    std::vector<int> rvar_bounds;
+    rvar_bounds.reserve(rvars.size());
+    for (int arg_id = 0; arg_id < (int)rvars.size(); arg_id++) {
+        Expr extent = simplify(substitute_var_estimates(rvars[arg_id].extent));
+        const int64_t *extent_int = as_const_int(extent);
+        user_assert(extent_int != nullptr) <<
+            "extent:" << extent << " is not constant.\n";
+        rvar_bounds.push_back(*extent_int);
+    }
+    return rvar_bounds;
+}
+
+void reorder_storage(Func func,
+                     const std::vector<Var> &all_vars,
+                     std::ostringstream &schedule_source) {
+    func.reorder_storage(all_vars);
+    schedule_source << "    .reorder_storage(";
+    for (int i = 0; i < (int)all_vars.size(); i++) {
+        schedule_source << all_vars[i].name();
+        if (i != (int)all_vars.size() - 1) {
+            schedule_source << ",";
+        }
+    }
+    schedule_source << ")\n";
+}
+
+void reorder_storage(Stage stage,
+                     const std::vector<Var> &all_vars,
+                     std::ostringstream &schedule_source) {
+    // Can't reorder storage of a stage.
+}
+
+int natural_vector_size(const Target &target, const Type &t) {
+    const bool is_integer = t.is_int() || t.is_uint();
+    const int data_size = t.bytes();
+    if (is_integer && (target.has_feature(Halide::Target::AVX512_Skylake) ||
+            target.has_feature(Halide::Target::AVX512_Cannonlake))) {
+        // AVX512BW exists on Skylake and Cannonlake
+        return 64 / data_size;
+    } else if (t.is_float() && (target.has_feature(Halide::Target::AVX512) ||
+            target.has_feature(Halide::Target::AVX512_KNL) ||
+            target.has_feature(Halide::Target::AVX512_Skylake) ||
+            target.has_feature(Halide::Target::AVX512_Cannonlake))) {
+        // AVX512F is on all AVX512 architectures
+        return 64 / data_size;
+    } else {
+        return 32 / data_size;
+    }
+}
+
+template <typename FuncOrStage>
+void parallelize_vars_and_rvars_gpu(
+        const MachineParams &params,
+        FuncOrStage func_or_stage,
+        bool is_pure_def,
+        const std::vector<Var> &vars,
+        const std::vector<int> &var_bounds,
+        const std::vector<RVar> &rvars,
+        const std::vector<int> &rvar_bounds,
+        TailStrategy tail,
+        std::ostringstream &schedule_source) {
+    // Find the first variable that has bounds larger or equal than 64,
+    // this is our GPU thread.
+    int split_size = 64;
```

Comment:
Why 64, what's special about it?
Also, here and elsewhere, split_size never seems to change, should it be const?

Reply:
It's 32*2 ; ) Made them constexpr.